packages/core/src/rules/common/assertions/utils.ts

```diff
@@ -0,0 +1,148 @@
+    const excludeParentKeys = node.excludeParentKeys;
+
+    if (matchParentKeys && excludeParentKeys) {
+      throw new Error(`Both 'matchParentKeys' and 'excludeParentKeys' can't be under one context item`);
+    }
+
+    if (matchParentKeys || excludeParentKeys) {
+      currentVisitorLevel[node.type] = {
+        skip: (_value: any, key: string) => {
+          if (matchParentKeys) {
+            return !matchParentKeys.includes(key);
+          }
+          if (excludeParentKeys) {
+            return excludeParentKeys.includes(key);
+          }
+        }
+      }
+    } else {
+      currentVisitorLevel[node.type] = {}
+    }
+    currentVisitorLevel = currentVisitorLevel[node.type];
+  }
+
+  currentVisitorLevel[subject] = subjectVisitor;
+
+  return visitor;
+}
+
+export function buildSubjectVisitor(properties: string | string[], asserts: Assert[], context?: Record<string, any>[]) {
+  return function(node: any, { report, location, key, type }: UserContext) {
+
+    // check context for same node type parent includes/excludes:
+    if (context) {
+      const lastContextNode = context[context.length-1];
+      if (lastContextNode.type === type.name) {
+        const matchParentKeys = lastContextNode.matchParentKeys;
+        const excludeParentKeys = lastContextNode.excludeParentKeys;
+
+        if (matchParentKeys && !matchParentKeys.includes(key)) {
+          return;
+        }
+        if (excludeParentKeys && excludeParentKeys.includes(key)) {
+          return;
+        }
+      }
+    }
+
+    if (properties) {
+      properties = Array.isArray(properties) ? properties : [properties];
+    }
+
+    for (const assert of asserts) {
+      if (properties) {
+        for (const property of properties) {
+          runAssertion(node[property], assert, location.child(property), report);
+        }
+      } else {
+        runAssertion(Object.keys(node), assert, location.key(), report);
+      }
+    }
+  }
+}
+
+export const getIntersectionLength = (keys: string[], properties: string[]): number => {
+  const props = new Set(properties);
+  let count = 0;
+  for (const key of keys) {
+    if (props.has(key)) {
+      count++;
+    }
+  }
+  return count;
+}
+
+export const isOrdered = (value: any[], options: OrderOptions | OrderDirection): boolean => {
+  const direction = (options as OrderOptions).direction || options as OrderDirection;
+  const property = (options as OrderOptions).property;
+  let result = true;
+  for (let i=1; i<value.length; i++) {
+    let currValue = value[i];
+    let prevVal = value[i-1];
+
+    if (property) {
+      if (!value[i][property] || !value[i-1][property]) {
+        return false; // property doesn't exist, so collection is not ordered
+      }
+      currValue = value[i][property];
+      prevVal = value[i-1][property]
+    }
+
+    result = direction === 'asc' ? currValue >= prevVal : currValue <= prevVal;
+    if (!result) {
+      break;
+    }
+  }
+  return result;
```

Comment:
This function is pretty easy to read though. I'm not sure if simplification is needed.

I would do some tiny change, instead of:

```diff
  if (!result) {
-   break;
+   return;
  }
```

and then


```diff
return result;
return true;
```

It makes it a little bit simpler to follow (in my opinion).